### DISCUSSION TITLE
BRouter integration - both or internal only?

### DISCUSSION THREAD

**Discussion by moving-bits** [cgeo/cgeo#10198] [Q&A] [answered]
While working on #10156 I've stumbled upon the following question:

Will c:geo-internal BRouter become the only routing option in c:geo, or shall we support both the internal variant and a separately installed BRouter app?

Supporting both would be relatively easy technically, at least for the time being, as I did not change anything in the interface between c:geo and BRouter. The only difference is that the service being called has a different name. But this is only in one place, so a config switch could easily take care of that.

Pro supporting both:
- no migration needed for existing users / no breaking change

Pro supporting internal only:
- slightly easier to maintain, no additional config option
- would give us the option to optimize the interface between BRouter and c:geo
- way easier for users using Android 11+ to use (configure) the automatic routing tile files downloader
- wizard could be simplified in regards to routing

Currently I see no urgent need for changes in the interfaces (which would be the main reason for me for option 2), and avoiding a harsh breaking change is something valuable, so I tend to favor the "both" option.

What do you think?

And follow-up question in case we should decide to use both: Which one do we support as default?

Proposal for that case: I would make using the internal BRouter the default for new installations, with automatic tile downloading being activated, so that getting to use routing within c:geo becomes very easy.

**Comment by SammysHP** [comment]
As long as I can use my already downloaded tiles that are used by other apps I don't care much. Although I prefer to not include BRouter in c:geo – I prefer more a KISS approach. Limit c:geo to Geocaching and interact with specialized apps like Locus Map or – since we already got Routing in c:geo – BRouter for route calculation.

**Reply by Lineflyer** [reply]
Well, theoretically I do agree...but the Brouter author seems to have no special interest to maintain it (at least I fear so from what he responded). So it might be wise to take it into c:geo before the feature is lost as the app might no longer be maintained / available.

**Comment by Lineflyer** [comment, marked as answer]
Reading your above posting with pro/con:

What about having a switch for internal/external, but in case the user wants external (pro=no need to change anything in his setting), let him do the maintenance (no c:geo internal tile downloading, etc.) So external 100% stays as it is today and internal provides the add-on feature (of tile download, etc.).
That would be my vote.

**Reply by moving-bits** [reply]
Agreed, sounds fair to me. I'll open a reminder issue for me, as implementing this in the current in-between-state of #10085 and #10156 does not make sense. I'll wait with this until `master` is reopened.

What about the second question - default value after a fresh install?

**Reply by Lineflyer** [reply]
> What about the second question - default value after a fresh install?

Puh...as long as Brouter is around on Google Play we have to expect users are already using it and we can just hop on.
Is it possible to set it to external if at installation time BRouter is already installed and only if not default to the internal?

This sounds good to me:
Users freshly starting with "geothings" or never had the competence to make use of routing in c:geo would get the internal version and have everything they need. Advanced Brouter users would get the external.
Once BRouter might vanish from Google Play we would automatically always have the internal set on installation (except for users still having it installed).

If this cannot be assigned dynamically, then internal would be the default for me as advanced users (those who have Brouter) would be probably clever enough to switch it to external.

**Reply by moving-bits** [reply]
A differentiation like "internal for new installations" and "external for existing users" would have to be programmed manually anyway, so your proposal "external if BRouter app is installed" and "internal otherwise" should not be a problem. I'll add that to #10203.

**Comment by fm-sys** [comment]
> What about the second question - default value after a fresh install?

I would vote for c:geo internal BRouter as that will be the one which is more user-friendly (no extra app, tile downloading...) or shall we make it depending on if external BRouter is installed?

EDIT: @Lineflyer was a little bit quicker but wrote pretty much the same 😅

**Reply by moving-bits** [reply]
Thanks to both of you :-)
